toroid.py

```diff
@@ -0,0 +1,304 @@
+        coefficient is always 1.
+        '''
+        x0 = ray_src[0]
+        y0 = ray_src[1]
+        z0 = ray_src[2]
+
+        ax = ray_dir[0]
+        ay = ray_dir[1]
+        az = ray_dir[2]
+
+        # Intermediate terms, from Graphics Gems
+        f = 1 - az*az
+        g = f + self.p*az*az
+        l = 2 * (x0*ax + y0*ay)
+        t = x0*x0 + y0*y0
+        q = self.A / (g*g)
+        m = (l + 2*self.p*z0*az) / g
+        u = (t + self.p*z0*z0 + self.B) / g
+        
+
+        # Final polynomial coeffs
+        c4 = 1
+        c3 = 2*m
+        c2 = m*m + 2*u - q*f
+        c1 = 2*m*u - q*l
+        c0 = u*u - q*t
+        return np.array([c4, c3, c2, c1, c0])
+
+
+    def ray_intersections(
+        self,
+        ray_src: Iterable[float],
+        ray_dir: Iterable[float],
+        quart_solver: Callable[[list[float]], (Iterable[float], Mapping)],
+    ) -> (list[float], dict):
+        '''Solves for intersection t values using the given quartic solver, and returns them in a
+        list alongside the locals from the solver.
+
+        Returns
+        -------
+        A list of distances along the given ray to its intersections, if any, with this toroid. 
+        This list is not guaranteed to be sorted.
+
+        Parameters
+        ----------
+        ray_src : Iterable[float] (length 3)
+            An array of length 3 corresponding to the x, y, and z coordinates of the ray's origin.
+        ray_dir : Iterable[float] (length 3)
+            An array of length 3 corresponding to the x, y, and z components of the ray's 
+            direction. This vector is assumed to be normalized to a magnitude of 1.
+        quart_solver : Callable[[list[float]], (Iterable[float], dict)]
+            A method which takes a quartic polynomial as a list of floats (ordered c4, c3 ... c0 
+            as returned by ~ray_intersection_polynomial), and returns its real roots, alongside
+            a copy of its local variables
+        '''
+        poly = self._ray_intersection_polynomial(ray_src, ray_dir)
+        t_vals, solver_locals = quart_solver(poly)
+        intersections: list = [t for t in t_vals if t > 0]
+        return intersections, solver_locals
+
+    def ray_intersection_points( 
+        self,
+        ray_src: np.array,
+        ray_dir: np.array,
+        quart_solver: Callable[[list[float]], (list[float], dict)],
+    ) -> tuple[list[np.array], Mapping]:
+        '''Solves for intersection points using the given quartic solver with ~ray_intersections, 
+        and returns them in a list alongside the locals from the solver.
+
+        Returns
+        -------
+        A list of the given ray's intersection points, if any, with this toroid.
+        As with ~ray_intersections, this list is not guaranteed to be sorted.
+
+        Parameters
+        ----------
+        ray_src : np.array (length 3)
+            An array of length 3 corresponding to the x, y, and z coordinates of the ray's origin.
```

Comment:
Here and through you don't need to repeat info about the type in the description

Reply:
i.e. you already say its a length 3 numpy array